redbot/cogs/mutes/mutes.py

```diff
@@ -0,0 +1,1212 @@
+        if requester != "discord_deleted_user":
+            return
+
+        await self._ready.wait()
+        all_members = await self.config.all_members()
+        for g_id, m_id in all_members.items():
+            if m_id == user_id:
+                await self.config.member_from_ids(g_id, m_id).clear()
+
+    async def initialize(self):
+        guild_data = await self.config.all_guilds()
+        for g_id, mutes in guild_data.items():
+            self._server_mutes[g_id] = {}
+            for user_id, mute in mutes["muted_users"].items():
+                self._server_mutes[g_id][int(user_id)] = mute
+        channel_data = await self.config.all_channels()
+        for c_id, mutes in channel_data.items():
+            self._channel_mutes[c_id] = {}
+            for user_id, mute in mutes["muted_users"].items():
+                self._channel_mutes[c_id][int(user_id)] = mute
+        self._ready.set()
+
+    async def cog_before_invoke(self, ctx: commands.Context):
+        await self._ready.wait()
+
+    def cog_unload(self):
+        self._unmute_task.cancel()
+        for task in self._unmute_tasks.values():
+            task.cancel()
+
+    async def is_allowed_by_hierarchy(
+        self, guild: discord.Guild, mod: discord.Member, user: discord.Member
+    ):
+        is_special = mod == guild.owner or await self.bot.is_owner(mod)
+        return mod.top_role.position > user.top_role.position or is_special
+
+    async def _handle_automatic_unmute(self):
+        await self.bot.wait_until_red_ready()
+        await self._ready.wait()
+        while True:
+            await self._clean_tasks()
+            try:
+                await self._handle_server_unmutes()
+            except Exception:
+                log.error("error checking server unmutes", exc_info=True)
+            await asyncio.sleep(0.1)
+            try:
+                await self._handle_channel_unmutes()
+            except Exception:
+                log.error("error checking channel unmutes", exc_info=True)
+            await asyncio.sleep(30)
+
+    async def _clean_tasks(self):
+        log.debug("Cleaning unmute tasks")
+        is_debug = log.getEffectiveLevel() <= logging.DEBUG
+        for task_id in list(self._unmute_tasks.keys()):
+            task = self._unmute_tasks[task_id]
+
+            if task.cancelled():
+                self._unmute_tasks.pop(task_id, None)
+                continue
+
+            if task.done():
+                try:
+                    r = task.result()
+                except Exception:
+                    if is_debug:
+                        log.exception("Dead task when trying to unmute")
+                self._unmute_tasks.pop(task_id, None)
+
+    async def _handle_server_unmutes(self):
+        log.debug("Checking server unmutes")
+        to_clear = []
+        for g_id, mutes in self._server_mutes.items():
+            to_remove = []
+            guild = self.bot.get_guild(g_id)
+            if guild is None:
```

Comment:
Cog disabling API needs to be handled here.
```suggestion
            if guild is None or await self.bot.cog_disabled_in_guild(self, guild):
```